Add linux memory fragmentation part two article (#572)
pingcap/blog: linux-kernel-vs-memory-fragmentation-2.md

---
title: Linux Kernel vs. Memory Fragmentation (Part II)
author: ['Wenbo Zhang']
date: 2021-05-03
summary: This post introduces the principle of memory compaction, how to view the fragmentation index, and how to quantify the latency overheads caused by memory compaction.
tags: ['Linux']
categories: ['Engineering']
image: /images/blog/linux-memory-fragmentation-and-defragmentation.png
---

> **Author:** Wenbo Zhang (Linux Kernel Engineer of the EE team at PingCAP)

Reviewer: ```suggestion
**Author:** [Wenbo Zhang](https://github.com/ethercflow) (Linux Kernel Engineer of the EE team at PingCAP)
```

**Transcreator:** [Charlotte Liu](https://github.com/CharLotteiu); **Editor:** Tom Dewan

![Linux kernel memory fragmentation and defragmentation](media/linux-memory-fragmentation-and-defragmentation.png)

> In [Linux Kernel vs. Memory Fragmentation (Part I)](https://en.pingcap.com/blog/linux-kernel-vs-memory-fragmentation-1), I concluded that grouping by migration types only delays memory fragmentation, but does not fundamentally solve it. As the memory fragmentation increases and it does not have enough contiguous physical memory, performance degrades.

Reviewer: ```suggestion
In [Linux Kernel vs. Memory Fragmentation (Part I)](https://pingcap.com/blog/linux-kernel-vs-memory-fragmentation-1), I concluded that grouping by migration types only delays memory fragmentation, but does not fundamentally solve it. As the memory fragmentation increases and it does not have enough contiguous physical memory, performance degrades.
```

Therefore, to mitigate the performance degradation, the Linux kernel community introduced **memory compaction** to the kernel.

In this post, I'll explain the principle of memory compaction, how to view the fragmentation index, and how to quantify the latency overheads caused by memory compaction.

## Memory compaction

> Before memory compaction, the kernel used lumpy reclaim for defragmentation. However, this feature was removed from v3.10 (currently the most widely used kernel version). If you’d like to learn more, you can read about lumpy reclaim in the articles I listed in [A brief history of defragmentation](https://en.pingcap.com/blog/linux-kernel-vs-memory-fragmentation-1#a-brief-history-of-defragmentation). For now, let me bring your mind to memory compaction.

Reviewer: ```suggestion
Before memory compaction, the kernel used lumpy reclaim for defragmentation. However, this feature was removed from v3.10 (currently the most widely used kernel version). If you'd like to learn more, you can read about lumpy reclaim in the articles I listed in [A brief history of defragmentation](https://pingcap.com/blog/linux-kernel-vs-memory-fragmentation-1#a-brief-history-of-defragmentation). For now, let me bring your mind to memory compaction.
```

Reviewer: Please update all quotation marks to straight and remove "en." from all URLs about the PingCAP website.

### Algorithm introduction

The article [Memory compaction](https://lwn.net/Articles/368869/) on LWN.net explains the algorithmic idea of memory compaction in detail. You can take the following fragmented zone as a simple example:

![A small fragmented memory zone](media/a-small-fragmented-memory-zone.png)
<div class="caption-center"> A small fragmented memory zone - LWN.net </div>

The white boxes are free pages, while those in red are allocated pages.

Memory compaction for this zone breaks down into three major steps:

1. Scan this zone from left to right for red pages of the MIGRATE_MOVABLE migration type.

    ![Search for movable pages](media/linux-kernel-movable-pages.png)

> 

Reviewer: ```suggestion
<div class="caption-center"> Search for movable pages </div>

```

Reviewer: To help SEO, we'd better add a caption for each image.

    ![The small memory zone after memory compaction](media/a-small-memory-zone-after-memory-compaction.png)

This principle seems relatively simple, and the kernel also provides `/proc/sys/vm/compact_memory` as the interface for manually triggering memory compaction.

However, as mentioned in [Part I](https://en.pingcap.com/blog/linux-kernel-vs-memory-fragmentation-1) and [Memory compaction issues](https://lwn.net/Articles/591998/), memory compaction is not very efficient in practice—at least for the most commonly-used kernel, v3.10—no matter whether it is triggered automatically or manually. Due to the high overhead it causes, it becomes a performance bottleneck instead.

The open source community did not abandon this feature but continued to optimize it in subsequent versions. For example, the community [introduced kcompactd](https://github.com/torvalds/linux/commit/698b1b3064) to the kernel in v4.6 and [made direct compaction more deterministic](https://lwn.net/Articles/686801/) in v4.8.

### When memory compaction is performed

In kernel v3.10, memory compaction is performed under any of the following situations:

> 

Reviewer: ```suggestion
```

The above steps are only a simplified description of the actual workflow. In real practice, it is more complicated and will be different depending on the requested memory order and allocation flags.

As for direct memory reclaim, it is not only performed by the kernel when the memory is severely insufficient, but also triggered due to memory fragmentation in practical scenarios. In a certain period, these two situations may occur simultaneously.

### How to analyze memory compaction

#### Quantify the performance latency

As mentioned in the previous section, the kernel may perform memory reclaim or memory compaction when allocating memory. To make it easier to quantify the latency caused by direct memory reclaim and memory compaction for each participating thread, I committed two tools, [drsnoop](https://github.com/iovisor/bcc/blob/master/tools/drsnoop_example.txt) and [compactsnoop](https://github.com/iovisor/bcc/blob/master/tools/compactsnoop_example.txt), to the [BCC](https://github.com/iovisor/bcc) project.

> Both tools are based on kernel events and come with detailed documentation, but there is one thing I want to note:  to reduce the cost of introducing Berkeley Packet Filters (BPF), these two tools capture the latency of each corresponding event. Therefore, you may see from the output that each memory request corresponds to multiple latency results.

Reviewer: ```suggestion
Both tools are based on kernel events and come with detailed documentation, but there is one thing I want to note: to reduce the cost of introducing Berkeley Packet Filters (BPF), these two tools capture the latency of each corresponding event. Therefore, you may see from the output that each memory request corresponds to multiple latency results.
```

The monitoring interfaces are simple to use, but they cannot perform quantitative analysis on the latency results, and the sampling period is too long. The tools based on kernel events can solve these problems, but you need a certain understanding of the working principle of the kernel-related subsystems to use these tools, and there are certain requirements for the client's kernel version.

Therefore, the monitoring interfaces and the kernel-events-based tools actually complement each other. Using them together can help you to analyze memory compaction thoroughly.

### How to mitigate memory fragmentation

The kernel is designed to take care of slow backend devices. For example, it implements the second chance method and the refault distance based on the LRU algorithm and does not support limiting the percentage of `page cache`. Some companies used to customize their own kernel to limit the `page cache` and tried to submit it to the upstream kernel community, but the community did not accept it. I think it may be because this feature causes problems such as working set refaults.

> Therefore,  to reduce the frequency of direct memory reclaim and mitigate fragmentation issues, it is a good choice to increase `vm.min_free_kbytes` (up to 5% of the total memory). This indirectly limits the percentage of `page cache` for scenarios with a lot of I/O operations, and the machine has more than 100 GB of memory.

Reviewer: ```suggestion
Therefore, to reduce the frequency of direct memory reclaim and mitigate fragmentation issues, it is a good choice to increase `vm.min_free_kbytes` (up to 5% of the total memory). This indirectly limits the percentage of `page cache` for scenarios with a lot of I/O operations, and the machine has more than 100 GB of memory.
```
